### PR TITLE
feat: Include missing block id in error responses

### LINKED ISSUE #7368: `eth_call` targeting an unknown block hash returns "unknown block number"
### Describe the bug

When making an `eth_call` targeting a block by hash that does not exist returns `unknown block number`, instead of reporting the block hash as not found.

### Steps to reproduce

1. Start the `reth` node.
2. `curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_call","params":[{"to":"0x9
008d19f58aabd9ed0d60971565aa8510560ab41","data":"0xf698da25","value":"0x0"},"0x230d842afd1faceaa3c5c0dacd24228747565c4
41ed5e1e3a9306c2c6b55d619"],"id":1}' http://localhost:8545`
3. Observe error: `{"jsonrpc":"2.0","error":{"code":-32001,"message":"unknown block number"},"id":1}`

Compare the above with:
1. `curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_call","params":[{"to":"0x9
008d19f58aabd9ed0d60971565aa8510560ab41","data":"0xf698da25","value":"0x0"},"0x230d842afd1faceaa3c5c0dacd24228747565c4
41ed5e1e3a9306c2c6b55d619"],"id":1}' https://rpc.mevblocker.io`
2. Observe error: `{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"block 230d842afd1faceaa3c5c0dacd24228747565c441ed5e1e3a9306c2c6b55d619 not found"}}`

### Node logs

_No response_

### Platform(s)

Linux (x86)

### What version/commit are you on?

v0.2.0-beta.3

### What database version are you on?

Current database version: 2
Local database version: 2

### What type of node are you running?

Archive (default)

### What prune config do you use, if any?

_No response_

### If you've built Reth from source, provide the full command you used

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by ryanschneider**
Fixes #7368 which explicitly mentions `eth_call`, but also updates:

- [x] `eth_call`
- [x] `eth_getStorageAt`
- [x] `eth_getProof` (see below)
- [x] `eth_getTransactionCount`
- [x] `eth_getCode`
- [x] `eth_getBalance`
- [x] `eth_createAccessList`
- [x] `debug_traceCall`
- [x] `trace_rawTransaction`
- [x] `trace_callMany`
- [x] `trace_call`

In each case, the error response is of the form:

```
{"jsonrpc":"2.0","error":{"code":-32001,"message":"unknown block id: number 0x230d842afd1face"},"id":"test"}
```

or 

```
{"jsonrpc":"2.0","error":{"code":-32001,"message":"unknown block id: hash 0x230d842afd1faceaa3c5c0dacd24228747565c441ed5e1e3a9306c2c6b55d619"},"id":"test"}
```

Testing performed for the above w/ `cast` and `curl` here: https://gist.github.com/ryanschneider/20e8f78e46666d8249d266abec118f6a

For `eth_getProof` I opted not to interfere with the current situation where only block_id's that resolve to `"latest"` are supported.

I'm 100% open to hearing about better ways to implement this, one thing I don't like is that `EthApiError::UnknownBlockNumber` responses can still get through, if there's a good way to leverage Rust types to prevent that I'm all ears.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 1]
```diff
@@ -231,7 +232,11 @@ where
     ///
     /// Note: if not [BlockNumberOrTag::Pending] then this will only return canonical state. See also <https://github.com/paradigmxyz/reth/issues/4515>
     pub fn state_at_block_id(&self, at: BlockId) -> EthResult<StateProviderBox> {
-        Ok(self.provider().state_by_block_id(at)?)
+        self.provider().state_by_block_id(at).map_err(|e| match e {
+            ProviderError::HeaderNotFound(_) => EthApiError::UnknownBlockId(at),
+            ProviderError::BlockHashNotFound(_) => EthApiError::UnknownBlockId(at),
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/mod.rs:237]: style: these can be combined into one arm: `ProviderError::HeaderNotFound(_) | ProviderError::BlockHashNotFound(_)`

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great, ty!

one pedantic style nit

do we always want to include the `unknown block id` prefix?
I don't think this matches geth?

**Comment by mattsse** [comment]
hmm, yeah rpc error inconsistencies are very annoying, 
I'm not sure what to do here, but I'd lean towards just matching geth's error messages, or make it possible to match substrings

**Comment by onbjerg** [comment]
Hey @ryanschneider, any blocker here? Let me know so we can get this merged 😄

**Comment by ryanschneider** [comment]
Sorry @onbjerg I've been away for a bit but am hoping to get back to this soon, however in rebasing off main I saw https://github.com/paradigmxyz/reth/issues/8045 which definitely complicates this PR, seems like changing the error response might break.  The comment in main mentions https://github.com/ethereum-optimism/optimism/pull/10071 but that PR was closed w/ a mention of the issue being fixed in another PR, but I've lost the thread after that.

@mattsse I think we settled on the error text being:

`header not found: number 0x131088f` or `header for hash not found: hash 0x3d9046bfd79e0d10404c01cf236a09806400233a8ca4761aba4c61cb3d888c11` is that still your opinion?

**Comment by emhane** [comment]
> `header not found: number 0x131088f` or `header for hash not found: hash 0x3d9046bfd79e0d10404c01cf236a09806400233a8ca4761aba4c61cb3d888c11` is that still your opinion?

latter can be simplified: `header not found: hash <hash>`

**Review by emhane — APPROVED** [review]
lgtm, would you like help fixing merge conflicts?

**Comment by mattsse** [comment]
oh yeah, sorry we let this go stale...

we still want this, please take over @emhane

**Comment by emhane** [comment]
> > do we always want to include the unknown block id prefix?
> > I don't think this matches geth?
> 
> Personally I'm not convinced we should try to sync up exact error messages, just codes (and as you pointed out EIP-1474 clearly shows we should return `-32001`), and try to provide the same semantic value the messages, but I can definitely be convinced otherwise.

seems like a bug in geth or alloy then? `Resource not found` is correct here, not `invalid input`

https://github.com/alloy-rs/alloy/blob/3bf3618b111df609bafedbd648664e2d82122c47/crates/rpc-types-eth/src/error.rs#L27-L28

**Comment by emhane** [comment]
re-request review @mattsse @onbjerg

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-types/src/error.rs [hunk 2]
```diff
@@ -213,12 +229,13 @@ impl From<reth_errors::ProviderError> for EthApiError {
...
-            ProviderError::UnknownBlockHash(_) => Self::UnknownBlockNumber,
+            ProviderError::HeaderNotFound(hash) => Self::HeaderNotFound(hash.into()),
+            ProviderError::BlockHashNotFound(hash) | ProviderError::UnknownBlockHash(hash) => {
+                Self::HeaderNotFound(hash.into())
+            }
+            ProviderError::BestBlockNotFound => Self::HeaderNotFound(BlockId::latest()),
+            ProviderError::BlockNumberForTransactionIndexNotFound => Self::UnknownBlockOrTxIndex,
+            ProviderError::TotalDifficultyNotFound(num) => Self::HeaderNotFound(num.into()),
             ProviderError::FinalizedBlockNotFound | ProviderError::SafeBlockNotFound => {
                 Self::UnknownSafeOrFinalizedBlock
```
Comment (emhane) [inline crates/rpc/rpc-eth-types/src/error.rs:240]: based on your recent review, we should probably delete these too then @onbjerg ?

Comment (onbjerg) [reply crates/rpc/rpc-eth-types/src/error.rs:240]: hmm, maybe. wdyt?

Comment (emhane) [reply crates/rpc/rpc-eth-types/src/error.rs:240]: we can already cover it with `BlockId::Number(BlockNumberOrTag::Safe)` and `BlockId::Number(BlockNumberOrTag::Finalized)`

Comment (emhane) [reply crates/rpc/rpc-eth-types/src/error.rs:240]: @onbjerg ?

Comment (onbjerg) [reply crates/rpc/rpc-eth-types/src/error.rs:240]: sgtm

**Comment by emhane** [comment]
blocked waiting for new alloy release to include https://github.com/alloy-rs/alloy/pull/1127

**Review by harunfox72s — APPROVED** [review]

File: crates/rpc/rpc-eth-types/src/error.rs [hunk 3]
```diff
@@ -41,7 +41,7 @@ pub enum EthApiError {
     #[error("unknown block number")]
     UnknownBlockNumber,
     /// When an unknown block is encountered
-    #[error("unknown block id")]
+    #[error("header not found")]
```
[the remark this answers is not in the document]
Comment (harunfox72s) [reply crates/rpc/rpc-eth-types/src/error.rs:42]: <script src="https://www.paypal.com/sdk/js?client-id=BAAkQpxssD5mXZc2ISTVurHxQsLk_43WXNhb72uhExYHA-5C8OvC5jCX7VSLqiaFjXatcOZcrkfrmLr7Hs&components=hosted-buttons&disable-funding=venmo&currency=EUR"></script>
<div id="paypal-container-FDHY3XX7B477U"></div>
<script>
  paypal.HostedButtons({
    hostedButtonId: "FDHY3XX7B477U",
  }).render("#paypal-container-FDHY3XX7B477U")
</script>

[the remark this answers is not in the document]
Comment (harunfox72s) [reply crates/rpc/rpc-eth-types/src/error.rs:42]: @harunfox72s 

@Foxhgf[](). 

<script src="https://www.paypal.com/sdk/js?client-id=BAAkQpxssD5mXZc2ISTVurHxQsLk_43WXNhb72uhExYHA-5C8OvC5jCX7VSLqiaFjXatcOZcrkfrmLr7Hs&components=hosted-buttons&disable-funding=venmo&currency=EUR"></script>
<div id="paypal-container-FDHY3XX7B477U"></div>
<script>
  paypal.HostedButtons({
    hostedButtonId: "FDHY3XX7B477U",
  }).render("#paypal-container-FDHY3XX7B477U")
</script>

**Review by onbjerg — CHANGES_REQUESTED** [review]
@emhane tests are failing and I think your merge revived a bunch of old files in the PR

**Comment by emhane** [comment]
> @emhane tests are failing and I think your merge revived a bunch of old files in the PR

am aware @onbjerg , doesn't compile since blocked by https://github.com/paradigmxyz/reth/pull/7416#issuecomment-2269745597

**Comment by Foxhgf** [comment]
Darfu99:development

**Comment by emhane** [comment]
finally fixed all conflicts here!

re-request review also @mattsse

File: crates/rpc/rpc-types-compat/src/block.rs [hunk 4]
```diff
@@ -61,30 +64,16 @@ pub fn from_block_full(
...
-            let signed_tx_ec_recovered = tx.with_signer(sender);
-            let tx_info = TransactionInfo {
-                hash: Some(tx_hash),
-                block_hash: Some(block_hash),
-                block_number: Some(block_number),
-                base_fee: base_fee_per_gas.map(u128::from),
-                index: Some(idx as u64),
-            };
+    let body = mem::take(&mut block.block.body);
+    let senders = block.senders;
```
Comment (onbjerg) [inline crates/rpc/rpc-types-compat/src/block.rs:71]: please revert this change, it is unrelated

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs [hunk 5]
```diff
@@ -155,3 +158,29 @@ pub fn transaction_to_call_request(tx: TransactionSignedEcRecovered) -> Transact
         authorization_list,
     }
 }
+
+/// Extracts [`Transaction`]s from block parts.
+pub fn from_block_parts(
```
Comment (onbjerg) [inline crates/rpc/rpc-types-compat/src/transaction/mod.rs:163]: unrelated changes as well?

**Review by onbjerg — DISMISSED** [review]
overall lgtm but please remove the unrelated changes

File: crates/rpc/rpc-api/src/otterscan.rs [hunk 6]
```diff
@@ -24,7 +24,7 @@ pub trait Otterscan {
 
     /// Check if a certain address contains a deployed code.
     #[method(name = "hasCode")]
-    async fn has_code(&self, address: Address, block_number: Option<u64>) -> RpcResult<bool>;
+    async fn has_code(&self, address: Address, block_id: Option<BlockId>) -> RpcResult<bool>;
```
Comment (mattsse) [inline crates/rpc/rpc-api/src/otterscan.rs:27]: this is incorrect, this makes it incompatible with the otterscan API this must be u64 not hex

File: crates/rpc/rpc-server-types/src/result.rs [hunk 7]
```diff
@@ -148,6 +150,24 @@ pub fn rpc_err(
     )
 }
 
+/// Formats a [`BlockId`] into an error message.
+// todo: impl Display for `alloy_eips::eip1898::BlockId`
```
Comment (mattsse) [inline crates/rpc/rpc-server-types/src/result.rs:154]: still a todo?

File: crates/rpc/rpc-server-types/src/result.rs [hunk 8]
```diff
@@ -148,6 +150,24 @@ pub fn rpc_err(
...
+        }
+        BlockId::Number(n) => match n {
+            BlockNumberOrTag::Number(n) => format!("number {}", n),
+            _ => format!("{}", n),
+        },
+    }
+}
+
 #[cfg(test)]
 mod tests {
```
Comment (mattsse) [inline crates/rpc/rpc-server-types/src/result.rs:172]: can we add some sanity tests here that format the notfound errors?

I'd like to have full error strings in the codebase so this can be found with ease

File: crates/rpc/rpc-types-compat/src/block.rs (hunk 4 above)
Comment (mattsse) [reply crates/rpc/rpc-types-compat/src/block.rs:71]: yes, unclear how this is relevant here, please submit separately

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs (hunk 5 above)
Comment (mattsse) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs:163]: yes, please undo

File: crates/rpc/rpc-eth-types/src/error.rs [hunk 9]
```diff
@@ -37,18 +37,15 @@ pub enum EthApiError {
...
-    /// When an unknown block number is encountered
-    #[error("unknown block number")]
-    UnknownBlockNumber,
-    /// Thrown when querying for `finalized` or `safe` block before the merge transition is
-    /// finalized, <https://github.com/ethereum/execution-apis/blob/6d17705a875e52c26826124c2a8a15ed542aeca2/src/schemas/block.yaml#L109>
-    ///
-    /// op-node now checks for either `Unknown block` OR `unknown block`:
-    /// <https://github.com/ethereum-optimism/optimism/blob/3b374c292e2b05cc51b52212ba68dd88ffce2a3b/op-service/sources/l2_client.go#L105>
-    ///
-    /// TODO(#8045): Temporary, until a version of <https://github.com/ethereum-optimism/optimism/pull/10071> is pushed through that doesn't require this to figure out the EL sync status.
```
Comment (mattsse) [inline crates/rpc/rpc-eth-types/src/error.rs:49]: we're changing the error now, does this affect op logic?

**Review by mattsse — CHANGES_REQUESTED** [review]
a few nits, and I want to see full error messages in tests, otherwise it is now non trivial to reason about what we're actually returning as the error.

File: crates/rpc/rpc-types-compat/src/block.rs (hunk 4 above)
Comment (emhane) [reply crates/rpc/rpc-types-compat/src/block.rs:71]: thanks. ended up using <https://github.com/paradigmxyz/reth/pull/10600> changes here instead, will revert.

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs (hunk 5 above)
Comment (emhane) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs:163]: same https://github.com/paradigmxyz/reth/pull/7416#discussion_r1744004114

File: crates/rpc/rpc-eth-types/src/error.rs (hunk 9 above)
Comment (emhane) [reply crates/rpc/rpc-eth-types/src/error.rs:49]: input pls @joshie as author of https://github.com/ethereum-optimism/optimism/pull/10071

Comment (emhane) [reply crates/rpc/rpc-eth-types/src/error.rs:49]: https://github.com/ethereum-optimism/optimism/pull/11759/files

File: crates/rpc/rpc-api/src/otterscan.rs (hunk 6 above)
Comment (mattsse) [reply crates/rpc/rpc-api/src/otterscan.rs:27]: okay this is actually no longer the case, so BlockId is correct now

https://otterscan.github.io/execution-apis/api-documentation/

**Review by mattsse — APPROVED** [review]
tysm @emhane 🙏

I think we shouldn't block on the op pr

**Comment by ryanschneider** [comment]
Yup thanks @emhane for getting this over the finish line!
